### PR TITLE
Fix: Failed to load model - unload model nitro

### LINKED ISSUE #572: bug: Failed to load model sometime
**Describe the bug**
Load model request is failed
Screenshot_2023-11-09_at_09 56 52

**To Reproduce**
- Start model, stop then restart, repeat this

### LINKED ISSUE #587: bug: Error occurred: Failed to fetch
**Describe the bug**
Cannot chat with Error: Failed to fetch

https://github.com/janhq/jan/assets/89722390/8b75ba43-8c80-47d7-998c-70ca4a00a3b1

https://github.com/janhq/jan/assets/89722390/74e332f1-707d-4b2e-9449-61bca219defb

**To Reproduce**
1. Go to 'chat'
2. Click on 'old conversations'

**Desktop (please complete the following information):**
 - Mac Air 0.3.1 Ventura
 - Version Jan 0.3.1

### PR DESCRIPTION
**Pull request by louis-jan**
## Problem
- The latest Nitro versions do not support runtime model reloading. This requires killing and respawning the process to work, but it's quite hacky. We should properly implement an unload feature.
- The main workflow is:
```
1. Check if the model file exists.
2. Check if the port is in use; if so, attempt to unload the model or kill the Nitro process.
3. Spawn the Nitro subprocess.
4. Wait until the port is in use (indicating the Nitro HTTP server is up).
5. Load the model into the Nitro subprocess via an HTTP POST request.
```

- This PR is also aimed at refactoring the entire inference module.ts, which has become difficult to maintain because all the logic is combined in the initModel function

```ts
function initModel(fileName: string): Promise<InitModelResponse> {
  // 1. Check if the model file exists
  return checkModelFileExist(fileName)
      // 2. Check if the port is used, if used, attempt to unload model / kill nitro process
      .then(checkAndUnloadNitro)
      // 3. Spawn the Nitro subprocess
      .then(spawnNitroProcess)
      // 4. Wait until the port is used (Nitro http server is up)
      .then(() => tcpPortUsed.waitUntilUsed(PORT, 300, 30000))
      // 5. Load the model into the Nitro subprocess (HTTP POST request)
      .then(() => loadLLMModel(fileName))
}
```

```
 * TODO: Should pass absolute of the model file instead of just the name - So we can modurize the module.ts to npm package (#618) ✅
 * TODO: Should it be startModel instead?
 * TODO: Should check model load status? (#618)✅
 * TODO: Should show Nitro running status somewhere?
```

Fixes #572 
Fixes #587 

#### NOTE: This PR is not complete due to a critical issue where Nitro 0.1.7 is incompatible with gguf v1 models, lacking backward compatibility. Therefore, two related updates are needed:
- Manage the handling of downloaded gguf v1 models and provide users with clearer messages and guidance.
- Update the model catalog to enable new users to download gguf v2 models

### CODE REVIEW

File: plugins/inference-plugin/src/module.ts [hunk 1]
```diff
@@ -1,119 +1,200 @@
...
+  const binaryPath = path.join(binaryFolder, binaryName);
+
+  // Execute the binary
+  subprocess = spawn(binaryPath, [1, "0.0.0.0", PORT], {
+    cwd: binaryFolder,
+  });
+
+  // Handle subprocess output
+  subprocess.stdout.on("data", (data) => {
+    console.log(`stdout: ${data}`);
```
[the remark this answers is not in the document]
Comment (louis-jan) [reply plugins/inference-plugin/src/module.ts:178]: @vuonghoainam, this task should only be handled by the Jan Log Manager (to be implemented later). Please refer to the 'jan-001-log-framework' JIP for more details.

File: plugins/inference-plugin/src/module.ts [hunk 2]
```diff
@@ -1,119 +1,200 @@
...
+  });
+
+  subprocess.stderr.on("data", (data) => {
+    console.error(`stderr: ${data}`);
```
[the remark this answers is not in the document]
Comment (louis-jan) [reply plugins/inference-plugin/src/module.ts:182]: @vuonghoainam, this task should only be handled by the Jan Log Manager (to be implemented later). Please refer to the 'jan-001-log-framework' JIP for more details.

**Comment by louis-jan** [comment]
@vuonghoainam, thank you for the information. Regarding the health check, it's the item I mentioned in the TODO list in the description. I'm not sure if it has been tested or what its response specifications are. Please provide details in a designated section and create a separate issue for this, as it will require a more complex retry mechanism, not just polling.

**Comment by louis-jan** [comment]
NOTE: This PR is not complete due to a critical issue where Nitro 0.1.7 is incompatible with gguf v1 models, lacking backward compatibility. Therefore, two related updates are needed:

Manage the handling of downloaded gguf v1 models and provide users with clearer messages and guidance.
Update the model catalog to enable new users to download gguf v2 models.

Waiting for #618 update & merge into this branch.

**Review by namchuai — APPROVED** [review]

**Pull request merged**